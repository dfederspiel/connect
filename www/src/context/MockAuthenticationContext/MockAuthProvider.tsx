import { useProvideMockAuth } from '.';
import { AuthProvider } from '../AuthenticationContext/';
import { AuthContext } from '../AuthenticationContext/types';
// import { AuthModule, Mode } from "./AuthModule";
/**
 * AuthContext to be passed to any component that depends on user context.
 */
// const authContext = createContext<AuthContext>(null as any);

// Provider component that wraps your app and makes auth object ...
// ... available to any child component that calls useAuth().
// eslint-disable-next-line react/prop-types

interface MockAuthProviderProps {
  children?: JSX.Element;
  user?: string;
}

export const MockAuthProvider = ({
  children,
  user,
}: MockAuthProviderProps): JSX.Element => {
  const auth = useProvideMockAuth(user) as AuthContext;
  return <AuthProvider provider={auth}>{children}</AuthProvider>;
};
