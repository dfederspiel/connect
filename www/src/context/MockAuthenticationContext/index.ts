export { MockAuthProvider } from './MockAuthProvider';
export { useProvideMockAuth } from './useProvideMockAuth';
