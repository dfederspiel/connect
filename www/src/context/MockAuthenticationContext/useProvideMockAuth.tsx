import { AuthContext } from '../AuthenticationContext/types';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface User {}

export const useProvideMockAuth = (name?: string): AuthContext => {
  const user = name;

  const signin = async (userPolicyId?: string): Promise<void> => {
    if (userPolicyId) sessionStorage.setItem('currentPolicy', userPolicyId || '');
  };

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const signout = (): void => {};

  const token = async (): Promise<string | null> => '123';

  return {
    user,
    signin,
    token,
    signout,
  };
};
