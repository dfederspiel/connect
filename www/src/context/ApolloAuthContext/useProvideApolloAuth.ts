import {
  ApolloClient,
  ApolloLink,
  FetchResult,
  HttpLink,
  InMemoryCache,
  Operation,
  split,
} from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import { getMainDefinition, Observable } from '@apollo/client/utilities';
import { OperationDefinitionNode, print } from 'graphql';
import { useAuth } from '../AuthenticationContext';
import { ApolloAuthContext } from './types';
import { createClient, ClientOptions, Client } from 'graphql-ws';

export const useProvideApolloAuth = (): ApolloAuthContext => {
  const { hostname, port, protocol } = window.location;
  const portString = port !== '' ? `:${port}` : port;
  const auth = useAuth();
  // const httpLink = createHttpLink({
  //   uri: `${process.env.APOLLO_HOST}`,
  // });
  const httpLink: ApolloLink = new HttpLink({
    uri: `${protocol}//${hostname}${portString}${process.env.APOLLO_HOST}`,
  });

  const authLink = setContext(async (_, { headers }) => {
    const token = await auth.token();
    console.log('GRAPHQL HTTP', token);
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : '',
      },
    };
  });

  // const wsLink = new WebSocketLink({
  //   uri: `${process.env.APOLLO_WS_HOST}`,
  //   options: {
  //     reconnect: true,
  //     connectionParams: async (): Promise<ConnectionParams> => {
  //       const token = await auth.token();
  //       console.log('GRAPHQL WS', token);
  //       return { authToken: token, test: 'some_other_thing' };
  //     },
  //   },
  // });

  class WebSocketLink extends ApolloLink {
    private client: Client;

    constructor(options: ClientOptions) {
      super();
      this.client = createClient(options);
    }

    public request(operation: Operation): Observable<FetchResult> {
      return new Observable((sink) => {
        return this.client.subscribe<FetchResult>(
          { ...operation, query: print(operation.query) },
          {
            next: sink.next.bind(sink),
            complete: sink.complete.bind(sink),
            error: sink.error.bind(sink),
          },
        );
      });
    }
  }
  console.log(
    // eslint-disable-next-line prettier/prettier
    `${protocol === 'https:' ? 'wss:' : 'ws:'}//${hostname}${portString}${process.env.APOLLO_HOST}`,
  );
  const wsLink = new WebSocketLink({
    // eslint-disable-next-line prettier/prettier
    url: `${protocol === 'https:' ? 'wss:' : 'ws:'}//${hostname}${portString}${process.env.APOLLO_HOST}`,
    connectionParams: async () => {
      const token = await auth.token();
      return { authToken: token, test: 'some_other_thing' };
    },
  });

  const link = split(
    ({ query }) => {
      const { kind, operation } = getMainDefinition(query) as OperationDefinitionNode;
      return kind === 'OperationDefinition' && operation === 'subscription';
    },
    wsLink,
    authLink.concat(httpLink),
  );

  const client = new ApolloClient({
    link,
    cache: new InMemoryCache(),
  });

  return {
    client,
  };
};
