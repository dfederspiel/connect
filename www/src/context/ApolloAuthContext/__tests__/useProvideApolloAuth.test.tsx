import { gql } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { MockAuthProvider } from '../../MockAuthenticationContext/MockAuthProvider';
import { renderHook } from '@testing-library/react-hooks';
import { useProvideApolloAuth } from '../useProvideApolloAuth';

beforeAll(() => {
  process.env.APOLLO_HOST = 'http://localhost/graphql';
  process.env.APOLLO_WS_HOST = 'ws://localhost/ws';
});

describe('the apollo auth context', () => {
  const wrapper = ({ children }: any) => (
    <MockAuthProvider user="user@contoso.com">
      <MockedProvider>{children}</MockedProvider>
    </MockAuthProvider>
  );

  it('renders with a client', () => {
    const { result } = renderHook(() => useProvideApolloAuth());
    expect(result.current.client).toBeDefined();
  });

  it('can query stuff', async () => {
    const mockFetchPromise = Promise.resolve({
      ok: true,
      status: 200,
      async text() {
        return JSON.stringify({
          data: {
            users: [],
          },
        });
      },
      async json() {
        return { ok: 'yay' };
      },
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    const { result } = renderHook(() => useProvideApolloAuth(), { wrapper });
    const response = await result.current.client.query({
      query: gql`
        {
          users {
            id
            email
          }
        }
      `,
    });
    expect(response).toBeDefined();
  });
});
