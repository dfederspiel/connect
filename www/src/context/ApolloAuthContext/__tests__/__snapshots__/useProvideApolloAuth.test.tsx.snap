// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the apollo auth context can query stuff 1`] = `
Object {
  "data": Object {
    "users": Array [],
  },
  "loading": false,
  "networkStatus": 7,
}
`;
